chapters/feature-selection.qmd

```diff
@@ -0,0 +1,525 @@
+
+Consider a hypothetical training set of 500 data points with a count-based predictor that has mostly zero counts. However, there are three data points with a count of 1 and one instance with a count of 2. The variance of these data is low, but it would be difficult to determine a cutpoint for removal. @kuhn2008building developed an algorithm to find **near-zero predictors**, as defined by two criteria:
+
+- The _freqeuncy ratio_ is the frequency of the most prevalent value over the second most frequent value. For our example, the ratio was 496/3 = 165.3.
+- The _proportion of unique values_, which is 3 / 500 = 0.006. 
+
+We can determine thresholds for both of these criteria to remove the predictor. The default behavior is to remove the column of the frequency ratio is greater than 19 **and** the proportion of unique values is less than 0.1. For the example, the predictor would be eliminated. 
+
+```{r}
+#| label: corr-data
+#| echo: false
+
+cor_data <- forested_train |>
+  select(-class)
+
+cor_mat <- cor_data %>%
+  cor(use = "pairwise.complete.obs")
+
+cut_0.50 <- cor_data %>%
+  recipes:::corr_filter(cutoff = .50) %>%
+  sort()
+
+lab_0.50 <- cli::format_inline(
+  "{xfun::numbers_to_words(length(cut_0.50))} predictors ({cut_0.50})"
+)
+```
+
+Another example of an unsupervised filter reduces between-predictor correlations. A high degree of correlation can reduce the effectiveness of some models. We can filter out predictors by choosing a threshold for the (absolute) pairwise correlations and finding the smallest subset of predictors to meet this criterion. 
+
+A good example is @fig-corr-plot, which visualizes the correlation matrix for the fifteen predictors in the forestry data set (from @sec-spatial-splitting). Predictors for the January minimum temperature and the annual mean temperature are highly correlated (corr = `r round(cor_mat["january minimum temperature", "annual mean temperature"], 2)`), and only one might be sufficient to produce a better model. Similar issues occur with other predictor pairs. If we apply a threshold of 0.50, then `r lab_0.50` would be removed before model training. 
+
+```{r}
+#| label: fig-corr-plot
+#| echo: false
+#| out-width: 90%
+#| fig-width: 5
+#| fig-height: 6
+#| fig-cap: A heatmap of a correlation matrix for fifteen predictors spatial, ordered using a hierarchical clustering algorithm.
+
+heatmaply(
+  cor_mat,
+  branches_lwd = .5,
+  colors = RdBu(50),
+  symm = TRUE,
+  revC = FALSE,
+  label_format_fun = function(...) round(..., digits = 2),
+  limits = c(-1, 1),
+  margins = c(50, 215, 10, 150)
+)
+```
+
+Similar redundancy can occur in categorical predictors. In this case, measures of similarity can be used to filter the predictors in the same way. 
+
+Unsupervised filters are only needed in specific circumstances. Some models (such as trees) are resistant to the distributions of the predictor’s data, while others are excessively sensitive. 
```

Comment:
Maybe add a note that while zero-variance filtering is only needed sometimes, it never hurts (and is VERY fast)